pages/DSNP/Graph.md

```diff
@@ -2,29 +2,124 @@
 
 This specification describes the social network graph and how it is represented in the protocol.
 
-In this context a [social graph](https://en.wikipedia.org/wiki/Social_graph) means a graph that represents social relations between entities.
+In this context a [social graph](https://en.wikipedia.org/wiki/Social_graph) means a graph that represents social relationships between entities.
 
-The DSNP graph represents nodes as [DSNP User Ids](Identifiers.md#dsnp-user-id).
+## Nodes
 
-## Public Graph
+Users are represented in the DSNP graph by nodes uniquely identified by their [DSNP User Ids](Identifiers.md#dsnp-user-id).
+Relationships between users are the graph edges.
 
-The public graph is anchored in state transitions recorded in [Graph Change Announcements](Types/GraphChange.md).
-Playing all state transitions in [order](Announcements.md#ordering-announcements) will generate the current state of the public graph.
+## Edges
 
-## Non-Normative
+Graph edges that originate from a user are stored as DSNP [User Data](UserData.md) associated with a given user.
 
-### Private Graph
+Each graph edge consists of the following data:
+* DSNP User Id (64-bit unsigned integer)
+  * To allow for optimal compression, User Ids are stored using the `long` type in Avro schema, which is a 64-bit _signed_ integer.
+    Care should be taken to ensure that User Id values greater than or equal to 2<sup>63</sup>, where used by an implementation, are converted correctly between signed and unsigned representations.
+* Timestamp when relationship was created (64-bit milliseconds since the [Unix epoch](https://en.wikipedia.org/wiki/Unix_time))
```

Comment:
My opinion is that it's useful to have, if only for the user themself to be able to track ("we've been friends since high school") or the app to use for curation ("happy friendiversary") etc. It also provides a useful way to give an ordering to the data that is a forcing function to put new relationships at the end, hopefully allowing for better transaction optimization.

As for granularity, I don't really see much actual use for more than a date, in which case 16 bits would be enough. If you add multiple friends on the same day, though, it does introduce some ambiguity from a consistent ordering perspective (you'd want to sort on date and then userId).

Reply:
A couple of further notes on this:
* Avro doesn't have a 16-byte primitive type, so we would have to define serialization to `fixed(2)`.
* For reference, the millisecond-based approach with an Avro `long` will ZigZag encode to 48 bits for the foreseeable future, so it's not as bad as the full 64 (but still a lot more than 16!)

Reply:
I'm on the side of seconds over milliseconds for data that usually goes on-chain.

Reply:
Re: `long` Our announcements have a Parquet data type. Do we also want to have an avro type in there?

Reply:
Agreed, changing to seconds.

Re: Parquet vs. Avro, can you clarify what you mean? I think Parquet is great for batches because you get native Bloom Filter support, self-contained schema metadata, and a columnar format that makes it easier to search. Those exact reasons make it a terrible choice for on-chain (or near-chain) data. As much as I'd rather support a single standard, the use cases suggest otherwise. If you mean we should optimize Parquet storage of INT64 types, it looks like that can be done with [DELTA_BINARY_PACKED = 5](https://parquet.apache.org/docs/file-format/data-pages/encodings/#a-namedeltaencadelta-encoding-delta_binary_packed--5). But if I'm misunderstanding the question let me know.

Reply:
I'm just thinking of adding the additional data type to the table data for each announcement. Aka instead of assuming that everything is parquet as it sort of does now, we are clearer in each Announcement spec how it should be represented in Avro data as well.

![image](https://user-images.githubusercontent.com/1252199/219405822-90deb834-02a8-4a40-9628-69f3d59c0b5a.png)

Reply:
OK I see what you mean. Let's break that out to a separate issue for discussion.

Reply:
See https://github.com/LibertyDSNP/spec/issues/235